src/Common/Scheduler/CPULeaseAllocation.h

```diff
@@ -0,0 +1,265 @@
+    const ResourceLink master_link; /// Resource link to use for master thread resource requests
+    const ResourceLink worker_link; /// Resource link to use for worker threads resource requests
+    const CPULeaseSettings settings;
+    LoggerPtr log;
+
+    /// Protects all the fields below
+    mutable std::mutex mutex;
+
+    /// Concurrency control (for interaction with consuming threads)
+    /// Every thread could be considered as a finite state machine w/states:
+    ///  * released: lease object was not created or was destructed, has no CPU slot
+    ///  * running: lease object owns a CPU slot
+    ///  * preempted: lease object does not own a CPU slot
+    /// Possible transitions:
+    ///  * released -> running: initial acquire() or tryAcquire() call
+    ///    - thread starts execution
+    ///  * running -> preempted: acquired slot was taken away during renew() and lease waits for another granted slot
+    ///    - thread execution is blocked
+    ///  * preempted -> running: newly granted slot was provided to lease
+    ///    - thread execution is unblocked
+    ///  * preempted -> released: timeout during renew() preemption wait - lease expired
+    ///    - renew() returns false and thread should stop itself
+    ///  * running -> released: lease destruction and release() of its acquired slot
+    ///    - thread execution stop voluntary (query is done/aborted/canceled)
+    struct Threads
+    {
+        explicit Threads(size_t max_threads_)
+            : leased(max_threads_)
+            , preempted(max_threads_)
+            , wake(max_threads_)
+        {}
+        boost::dynamic_bitset<> leased; /// Thread lease object status bitmask (0=released; 1=preempted|running)
+        boost::dynamic_bitset<> preempted; /// Preempted threads bitmask (0=running|released; 1=preempted)
+        std::vector<std::condition_variable> wake; /// To wake specific preempted thread
+
+        // For optimization (could be computed based on leased and preempted fields)
+        size_t running_count = 0; /// Number of currently running threads (leased & !preempted)
+        size_t last_running = boost::dynamic_bitset<>::npos; /// Highest thread num of a running threads
+    } threads;
+
+    /// Resource accounting
+    std::atomic_bool acquirable{false}; // Tracks `(granted > 0 || exception) && !shutdown` value that could be read w/o locking mutex
+    SlotCount allocated = 0; /// Current number of allocated (granted and acquired) slots
+    Int64 granted = 0; /// Allocated but not acquired slots (might be negative if acquired more than allocated)
+    ResourceCost consumed_ns = 0; /// Real consumption accumulated from renew() calls
+    ResourceCost requested_ns = 0; /// Consumption requested from the scheduler (requested <= consumed + quantum)
+
+    /// Scheduling control (for interaction with resource scheduler)
+    using Requests = std::vector<Request>;
+    Requests requests; /// Circular buffer of requests per every slot
```

Comment:
Can we introduce small abstraction on top of vector of requests and head + tail. Right now all these `++tail`, `--head` and `tail == requests.end()` is very confusing.

Reply:
Fixed. Now we have the `RequestChain` class to deal with resource requests.